### DISCUSSION TITLE
How to use isDirty in window.onbeforeunload?

### DISCUSSION THREAD

**Discussion by PraveenPeri** [react-hook-form/react-hook-form#5339] [Q&A]
This is more of a question than a bug report. How can I use formState's isDirty to warn users when closing the window?

1. Read the formState before render like `const { formState: { isDirty } } = useForm()`
2. Add event handler to the window.onbeforeUnload in the useEffect <br/>
`useEffect(() => {
    window.addEventListener("beforeunload", () => {
        if(isDirty){
            return "You have unsaved changes in the form. Save them before exiting"
        }
    });
});`

**Expected behavior**
The window displays a warning before closing

**Actual behavior**
The window closes without any warning

**Comment by bluebill1049** [comment]
provide a codesandbox.

**Reply by PraveenPeri** [reply]
https://codesandbox.io/s/magical-galois-9fwds

**Reply by bluebill1049** [reply]
<img width="1112" alt="Screen Shot 2021-05-26 at 2 56 11 pm" src="https://user-images.githubusercontent.com/10513364/119604502-84c3d180-be32-11eb-938e-97e39c2ccaea.png">

it should work, right? formState is updated correctly. So in your example, `isDirty` remain `false`?

**Reply by PraveenPeri** [reply]
The isDirty doesn't seem to carry into the event handler.

**Reply by bluebill1049** [reply]
Try with useCallback with formState as dep

**Reply by PraveenPeri** [reply]
No, it didn't work! Also, when I added a keypress event handler to the input field that logs isDirty value in the console, it logs the value as false and then true for every keypress
![image](https://user-images.githubusercontent.com/6571000/119682520-cd11ce00-bdf7-11eb-8325-a41fb103c755.png)

**Reply by PraveenPeri** [reply]
@bluebill1049 Any update on this?

**Comment by leapful** [comment]
@PraveenPeri @bluebill1049 

Actually, each render of React component will have its own props, states and effect. The value of "formState.isDirty" inside "useEffect" or  "event handler" will be captured at the same time of its rendering.

In the CSB sample, "useEffect" is used without putting dependency "isDirty", meaning this "useEffect" will be perform on each rendering, but then there is no cleanup handling for event "beforeunload", therefore, the first old "even handler" is still being used, that's why "isDirty" is always false.

It should be something like:

```

useEffect(() => {

   const onBeforeUnload = () => {
      if (isDirty) { // handle before unload message }
   }

  window.addEventListener('beforeunload', onBeforeUnload);

  return () => {
     window.removeEventListener('beforeunload', onBeforeUnload);
  }
}, [isDirty]);

```

Working sample: https://codesandbox.io/s/cranky-wave-pf0lq?file=/src/App.js

P.S: If you want to avoid perform adding and removing event listener every time the dirty state is changed, you can use "useRef" to keep "isDirty" value.

For example:
```

const { formState: { isDirty } } = useForm();

const dirty = React.useRef(false);
dirty.current = isDirty;

React.useEffect(() => {
   const onBeforeUnload = () => {
     if (dirty.current) { }
   }
   window.addEventListener('beforeunload', onBeforeUnload);

   return () => {
      window.removeEventListener('beforeunload', onBeforeUnload);
   }
}, []);

```

**Reply by PraveenPeri** [reply]
Thank you @leapful! Appreciate your response! I will give this a try and let you know.

**Reply by bluebill1049** [reply]
Thank you @leapful

**Reply by jakewisse** [reply]
What's the need for the `dirty` ref above? Can't you just reference `formState.isDirty` returned from `useForm()`?

**Reply by jakewisse** [reply]
Nvm, `isDirty` would take its value from the closure of when the effect was run, not when the listener is invoked. Gosh that looks weird though.

**Comment by yannickguillemot** [comment]
Following up on this discussion after recollecting conflicting information....

@bluebill1049 first solution was almost accurate. The way to activate the "Changes not saved" dialog is different for modern browsers than your example.  
```js
if (isDirty) {
  e.preventDefault();
  e.returnValue = "Save your changes before exiting.";
  // return "Save your changes before exiting."; -- Old method
}
```

Full example: https://codesandbox.io/s/wild-tdd-4gchou?file=/src/App.js

<img width="1786" alt="CleanShot 2022-12-05 at 16 16 44@2x" src="https://user-images.githubusercontent.com/1564064/205744451-aadee6b8-3037-4a3b-a48d-555d073f93dc.png">
